java-checks/src/main/java/org/sonar/java/checks/DateTimeFormatterMismatchCheck.java

```diff
@@ -0,0 +1,187 @@
+    .names("appendValue")
+    .addParametersMatcher(TEMPORAL_FIELD_TYPE)
+    .addParametersMatcher(TEMPORAL_FIELD_TYPE, INT_TYPE)
+    .addParametersMatcher(TEMPORAL_FIELD_TYPE, INT_TYPE, INT_TYPE, "java.time.format.SignStyle")
+    .build();
+
+  private static final String CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE = "Change this year format to use the week-based year instead" +
+    " (or the week format to Chronofield.ALIGNED_WEEK_OF_YEAR).";
+  private static final String CHANGE_YEAR_FORMAT_TO_CHRONOFIELD_MESSAGE = "Change this year format to use ChronoField.YEAR instead" +
+    " (or the week format to WeekFields.ISO.weekOfWeekBasedYear()).";
+  private static final String SECONDARY_LOCATION_MESSAGE = "Week format is inconsistent with year.";
+
+  private int depth = 0;
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree invocation = (MethodInvocationTree) tree;
+    if (APPEND_VALUE_MATCHER.matches(invocation)) {
+      if (depth == 0) {
+        ChainedInvocationVisitor visitor = new ChainedInvocationVisitor();
+        invocation.accept(visitor);
+        if (visitor.usesWeek && visitor.usesYear) {
+          if (visitor.usesWeekBasedYear && !visitor.usesWeekOfWeekBasedYear) {
+            reportIssue(visitor.primary, CHANGE_YEAR_FORMAT_TO_CHRONOFIELD_MESSAGE, visitor.secondaries, null);
+          } else if (!visitor.usesWeekBasedYear && visitor.usesWeekOfWeekBasedYear) {
+            reportIssue(visitor.primary, CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE, visitor.secondaries, null);
+          }
+        }
+      }
+      depth++;
+    }
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
+    if (APPEND_VALUE_MATCHER.matches((MethodInvocationTree) tree)) {
+      depth--;
+    }
+  }
+
+  @Override
+  public void leaveFile(JavaFileScannerContext context) {
+    depth = 0;
+  }
+
+
+  private static class ChainedInvocationVisitor extends BaseTreeVisitor {
+    private static final MethodMatchers WEEK_BASED_YEAR_MATCHER = MethodMatchers.create()
+      .ofTypes("java.time.temporal.WeekFields")
+      .names("weekBasedYear")
+      .addWithoutParametersMatcher()
+      .build();
+
+    private static final MethodMatchers WEEK_OF_WEEK_BASED_YEAR_MATCHER = MethodMatchers.create()
+      .ofTypes("java.time.temporal.WeekFields")
+      .names("weekOfWeekBasedYear")
+      .addWithoutParametersMatcher()
+      .build();
+
+    private boolean usesWeek = false;
+    private boolean usesWeekOfWeekBasedYear = false;
+    private boolean usesYear = false;
+    private boolean usesWeekBasedYear = false;
+    private final List<JavaFileScannerContext.Location> secondaries = new ArrayList<>();
+    private ExpressionTree primary = null;
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree callToAppendValue) {
+      MethodInvocationTree tree = callToAppendValue;
```

Comment:
This variable is redundant. I'd rename the parameter name to `tree` and remove this assignment.